### DISCUSSION TITLE
installing PEST on Linux (and macOS)

### DISCUSSION THREAD

**Discussion by JackElsey** [gmdsi/GMDSI_notebooks#92] [General]
The [PEST Basics](https://github.com/gmdsi/GMDSI_notebooks/tree/main/tutorials/part1_02_pest_setup) tutorial in Part 1 assumes that you have PEST installed on your system. Here are some instructions for doing that if you're on Linux. These instructions should also work if you're using macOS.

**Step 1**
Download the source code for PEST by visiting the project's [Programs](https://pesthomepage.org/programs) page and clicking "Download PEST source code to compile for Linux".

**Step 2**
After extracting the downloaded file, you should have a directory called `pest_source`. Move this directory to where you would like to keep the PEST executable files (perhaps the folder containing the GMDSI tutorials).

**Step 3**
PEST is primarily written in Fortran code, which must be compiled before running. To compile Fortran, you need a program called `gfortran`. Install it by entering `sudo apt install gfortran` into a terminal.

**Step 4**
Programs are compiled using `make` commands. Full instructions for compiling PEST are given in `pest_source/doc/unixpest.pdf`. Here's a script for automating that compilation process. Copy the text below into a file you save as `make_pest.sh` in the directory `pest_source`.

```bash
#!/bin/bash

make cppp
make -f pest.mak all
make clean
make -f ppest.mak all
make clean
make -f pestutl1.mak all
make clean
make -f pestutl2.mak all
make clean
make -f pestutl3.mak all
make clean
make -f pestutl4.mak all
make clean
make -f pestutl5.mak all
make clean
make -f pestutl6.mak all
make clean
make -f pestutl7.mak all
make clean
make -f sensan.mak all
make clean
make –f beopest.mak all
make clean
make install
```

Then make that bash script executable by entering `chmod u+x make_pest.sh` into your terminal. Run the bash script via `./make_pest.sh`

**Step 5**
If the compilation was successful, you should have ~170 new executable files (e.g. `tempchek` and `inschek`) in `pest_source`. These executable files are run in the terminal just like the bash script we made above. For example, this is the command I used when the [PEST Basics](https://github.com/gmdsi/GMDSI_notebooks/tree/main/tutorials/part1_02_pest_setup) tutorial asked me to check one of the modified-Freyberg model template files.

```bash
~/Documents/Study/pest_source/tempchek pest_files/freyberg6.npf_k_layer1.txt.tpl
```

**Comment by jtwhite79** [comment]
Thanks @JackElsey .  I think all we are missing from the linux side is `tempcheck` and `inschek` and they only get used in that one notebook.  Otherwise, we only use PEST++ binaries in the notebooks and we have added precompiled binaries for all three OS's to the repo and they (should) get copied into the current working dir of each notebook programmatically by `prep_deps()`.  

@rhugman - I think we can ditch the `tempchek` and `inschek` by using `check_tplins True` and `debug_parse_only True` with any of the pest++ bins.  Is that ok with you?  Would help reduce the number of bins needed...

**Reply by JackElsey** [reply]
Tutorial [1.3: Automated Calibration with PEST](https://github.com/gmdsi/GMDSI_notebooks/tree/main/tutorials/part1_03_calibrate_k) also requires binaries of `tempchek`, `inschek`, and `pestchek`.

**Comment by laat0003** [comment]
Thanks for posting this @JackElsey. I didn't need it for the notebooks because I'm on Windows but I've started playing around with using Linux for modelling purposes and this information was very helpful.
